### PR TITLE
Comm node

### PR DESCRIPTION
**Pull request by andrew103**
Implement ROS node as an interface to the serial port and a node to test the interface with random data. Also includes empty `comm_socket` file that's unused currently

### CODE REVIEW

**Comment by andrew103** [comment]
Note: testing node doesn't explicitly try edge cases

File: trajectory_executor/com_serial.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_R_ELBOW']
+        elif read_state == serial_read_states['READ_R_ELBOW']:
+          data_packet['r_elbow'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_AT_GOAL']
```
Comment (goobta) [inline trajectory_executor/com_serial.py:155]: Could be replaced with something like this?
```python
for cmd, state in serial_read_states.items():
  if read_state == state:
    data_packet[cmd[5:].lower()] += decoded_data
    calculated_checksum += decoded_data
    data_byte_counter -= 1

    if data_byte_counter == 0:
      data_byte_counter = self.data_byte_length
      read_state = serial_read_states[serial_read_states.index(cmd) + 1]

    break
```

idk about the preamble and the checksum, but I believe that should encompass all of the logic for the sensor readings.

File: trajectory_executor/com_serial.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            response.yaw = data_packet['yaw']
+            response.l_hip = data_packet['l_hip']
+            response.l_knee = data_packet['l_hip']
+            response.r_hip = data_packet['r_hip']
+            response.r_knee = data_packet['r_hip']
+            response.l_shoulder = data_packet['l_shoulder']
+            response.l_elbow = data_packet['l_elbow']
+            response.r_shoulder = data_packet['r_shoulder']
+            response.r_elbow = data_packet['r_elbow']
+            response.at_goal = data_packet['at_goal']
```
Comment (goobta) [inline trajectory_executor/com_serial.py:175]: How does ROS handle these automatically generated `Request` objects? Basically, can you instantiate it and set the values at construction time--ie. `response = SensorDataRequest(roll=<roll>, pitch=<pitch>, ...)`?

If so, then you can use keyword expansion and do all of this in one swoop with 
```python
response = SensorDataRequest(**data_packet)
```

File: trajectory_executor/com_serial.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            break
+          else:
+            # bad packet, retry request
+            return self.sensor_data_request_callback(request, response)
```
Comment (goobta) [inline trajectory_executor/com_serial.py:179]: Would this default behavior cause any timing issues? Particularly, I'm just concerned if a packet fails and in retrying, we end up backing the buffer. Again, not an expert by any means so if that's not how things work, lmk!

File: trajectory_executor/com_serial.py (hunk 2 above)
Comment (goobta) [reply trajectory_executor/com_serial.py:175]: can also do
```python
for attr, value in data_packet.items():
   setattr(response, attr, value)
```
with the assumption that the keys in `data_packet` correspond to the attributes in `response`

File: trajectory_executor/com_serial.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    ja_bytes.append(255 if msg.right_knee // 256 > 0 else msg.right_knee)
+    ja_bytes.append(msg.right_knee % 255 if msg.right_knee // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.left_shoulder // 256 > 0 else msg.left_shoulder)
+    ja_bytes.append(msg.left_shoulder % 255 if msg.left_shoulder // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.left_elbow // 256 > 0 else msg.left_elbow)
+    ja_bytes.append(msg.left_elbow % 255 if msg.left_elbow // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.right_shoulder // 256 > 0 else msg.right_shoulder)
+    ja_bytes.append(msg.right_shoulder % 255 if msg.right_shoulder // 256 > 0 else 0)
+    ja_bytes.append(255 if msg.right_elbow // 256 > 0 else msg.right_elbow)
+    ja_bytes.append(msg.right_elbow % 255 if msg.right_elbow // 256 > 0 else 0)
```
Comment (goobta) [inline trajectory_executor/com_serial.py:202]: again, not 100% sure on how ROS creates its data objects, but this could potentially be reduced to:

```python
joints = (attr for attr in dir(msg) if not attr.startswith('_'))
ja_bytes = []
for joint in joints:
  if getattr(msg, joint) // 256 > 0:
    ja_bytes.append(255)
    ja_bytes.append(getattr(msg, joint) % 255)
  else:
    ja_bytes.append(getattr(msg, joint))
    ja_bytes.append(0)
```

File: trajectory_executor/com_serial.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    )
+
+  def sensor_data_request_callback(self, request: SensorDataRequest, response: SensorDataRequest):
+    """Ping Arduino for sensor data"""
+    request_sensors = 1
+    checksum = 255 - (request_sensors * self.data_byte_length) % 256
+
+    # send request for sensor data
+    for i in range(self.preamble_length):
+      self.arduino_port.write(struct.pack('>B', 255))
```
Comment (goobta) [inline trajectory_executor/com_serial.py:49]: noticed that `'>B'` is used a lot. is it a standard? either way feels like a magic string, maybe we should have a comment or make it into a named variable

File: trajectory_executor/com_serial.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+    # calculate checksum
+    checksum = 255 - sum(ja_bytes) % 256
+    ja_bytes.append(checksum)
+
+    # send preamble, joint angles, and checksum
+    for i in range(self.preamble_length):
+      self.arduino_port.write(struct.pack('>B', 255))
+    for ja_byte in ja_bytes:
+      self.arduino_port.write(struct.pack('>B', ja_byte))
```
Comment (goobta) [inline trajectory_executor/com_serial.py:214]: since we only need to be returning joint angles right now, hardcoding the preamble and checksum sandwiching is probably fine. in the future, this should probably be abstracted out cause it's required for all communications.

File: trajectory_executor/datastructures.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  'l_knee': 0,
+  'r_hip': 0,
+  'r_knee': 0,
+  'l_shoulder': 0,
+  'l_elbow': 0,
+  'r_shoulder': 0,
+  'r_elbow': 0,
+  'at_goal': 0,
+  'checksum': 0,
+}
```
Comment (goobta) [inline trajectory_executor/datastructures.py:33]: Will these two dictionaries always be repeated? if so, could they be merged by perhaps making a richer data structure such as a `collections.named_tuple` or a similar data structure?

**Review by goobta — COMMENTED** [review]
Comments attached. Also, do we have tests for the logic? Testing serial ports is probably more hassle than its worth, but there's a lot of logic in `com_socket` that is _super_ susceptible to off-by-one errors--it'd be nice to have the piece of mind that it all works :)

As always, gonna let @mahajanrevant give the final approval because he actually knows what's going on.

**Comment by andrew103** [comment]
> Comments attached. Also, do we have tests for the logic? Testing serial ports is probably more hassle than its worth, but there's a lot of logic in `com_socket` that is _super_ susceptible to off-by-one errors--it'd be nice to have the piece of mind that it all works :)
> 
> As always, gonna let @mahajanrevant give the final approval because he actually knows what's going on.

Similar to the arduino setup, there aren't any tests that explicitly check the logic other than looking at the output on the other end. I could setup the testing node with a fixed set of values to send instead of random ones to be sure everything is correct.

File: trajectory_executor/com_serial.py (hunk 1 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:155]: I'm a fan of this idea. Will look into implementing it.

File: trajectory_executor/com_serial.py (hunk 2 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:175]: I'm not sure. In the tutorials, they've always created the object first then populated fields. I can post a question on the ROS answers forum and experiment with it a little.

File: trajectory_executor/com_serial.py (hunk 3 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:179]: Yea I understand. I was skeptical even when adding this in. It's actually not fully implemented. My thought process for this was to have a number of attempts that it tracked and bails out if it exceeds say 5 tries. Might be better to just bail out immediately considering we'll probably be running the service call pretty frequently.

File: trajectory_executor/com_serial.py (hunk 4 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:202]: Yea so this one I did look into a bit because it's really ugly and I hate it but so far it seems that there isn't a way to iterate over the items of a ROS message like you would do with a python dictionary. One way to get around this would be to have the message object only have one field: a list of numbers representing the joint angles. I avoided this solution for the first run just because we lose explicit assignment of values to their respective joints and rely on the list order being maintained during transmission. Purely documentation reasons. I'm ok with switching the message format to a list if you guys are ok with it and really want to get rid of this unholy mess.

File: trajectory_executor/com_serial.py (hunk 5 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:49]: I learned about this when working on my previous robot. The `'>B'` tells `struct.pack` how to encode the data (in this case encode into bytes). There might be a better/cleaner way to encode the data as bytes, but this is the first solution I found a while back for my other robot and I just stuck with it. We could make it a named variable for organization.

File: trajectory_executor/com_serial.py (hunk 6 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:214]: I suppose. Although, I'm in the boat of only abstract if its necessary, and I don't see it as necessary here where there's only two instances of each operation. Sure there's repetition but it's only a few lines so I don't see the need. I we were gonna add more modes of communication that needed to also send preambles/checksums then I would be all for it, but we're very likely only using the current two communication channels.

File: trajectory_executor/datastructures.py (hunk 7 above)
Comment (andrew103) [reply trajectory_executor/datastructures.py:33]: Could you elaborate? I'm not sure I follow. Are you saying to put these two dictionaries together as part of a parent dictionary? i.e.
```
collections = {
    'serial_read_states': {...},
    'data_packet_struct': {...}
}
```

File: trajectory_executor/com_serial.py (hunk 2 above)
Comment (goobta) [reply trajectory_executor/com_serial.py:175]: iirc the second example should work cause I believe that they are just Python objects. idk feel free to experiment with it. I'm not sure how much the data packet will change down the line but this would make it fully dynamic.

File: trajectory_executor/datastructures.py (hunk 7 above)
Comment (goobta) [reply trajectory_executor/datastructures.py:33]: You could try something like:

```
_State = namedtuple('State', ['serial', 'data_name', 'order'])
```
and create the dict by something like 
```
states = {
  'roll' = _State('READ_ROLL', 'roll', 2),
}
```
but the more I think about it, that might just be overkill. idk, whatever works!

Comment (andrew103) [reply trajectory_executor/datastructures.py:33]: Yea in this case I'll elect to keep it as is for the sake of simplicity. Later down the line if there's time and want, we can try changing it up.

File: trajectory_executor/com_serial.py (hunk 5 above)
Comment (mahajanrevant) [reply trajectory_executor/com_serial.py:49]: Add this as a comment so it's understandable. For the future, consider adding comments to things that might not be so apparent.

File: trajectory_executor/com_serial.py (hunk 1 above)
Comment (mahajanrevant) [reply trajectory_executor/com_serial.py:155]: I was looking into this for ros2sim. When I implement that, we can just use it here. So don't worry about this

File: trajectory_executor/com_serial.py (hunk 2 above)
Comment (mahajanrevant) [reply trajectory_executor/com_serial.py:175]: I think we can test this out with a simple service example from ROS. I like the idea of it being dynamic

File: trajectory_executor/com_serial.py (hunk 3 above)
Comment (mahajanrevant) [reply trajectory_executor/com_serial.py:179]: I think we should log that we received a bad packet. Also let's have a limit on retry like you suggested Andrew. We should define retry behavior for both ros-sim and ros-realrobot. These should be identical so it would be best if we document that. **Maybe just add the logging here itself?**

File: trajectory_executor/com_serial.py (hunk 4 above)
Comment (mahajanrevant) [reply trajectory_executor/com_serial.py:202]: I hate this holy mess but I think we have other important priorities. I am okay with creating a bunch of code improvement related issues so that we can resolve them. I don't think it's wise to spend time on this right this moment

File: trajectory_executor/com_serial.py (hunk 6 above)
Comment (mahajanrevant) [reply trajectory_executor/com_serial.py:214]: Again let's add this as a code improvement issue with all the necessary details and hash it out later

**Review by mahajanrevant — COMMENTED** [review]
Some comments for you to look at @andrew103.
@agupta231 give them a glance too

File: trajectory_executor/com_serial.py (hunk 1 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:155]: Sounds good.

File: trajectory_executor/com_serial.py (hunk 2 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:175]: Yea there are ways to do various testing on it. For now let's leave it as is and make a ticket to revisit this

File: trajectory_executor/com_serial.py (hunk 3 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:179]: Yea we can just log it from here. Regarding the retries, after thinking about it more I think it would be better to just have it bail out rather than block the process to retry. We'll be calling the service frequently enough that some errors shouldn't be an issue. Even then, I doubt we'll have a significant number of errors.

File: trajectory_executor/com_serial.py (hunk 4 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:202]: Yep. We'll create a ticket in trello for this

File: trajectory_executor/com_serial.py (hunk 5 above)
Comment (andrew103) [reply trajectory_executor/com_serial.py:49]: Updated with added comment

**Review by mahajanrevant — APPROVED** [review]
Looks good!! I am going to approve it from my end. I have resolved the conversations but @agupta231 feel free to look at them. I am leaving this to you to merge it in

**Comment by goobta** [comment]
lgtm, we can address everything else in future PRs.
